Ex6/Ex6.java

```diff
@@ -0,0 +1,118 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.HashMap;
+
+public class Ex6 {
+	private int availableSpot;
+	private int carToMove;
+	private int[] initial;
+	private int[] desired;
+	private ArrayList<Move> moves;
+	
+	public Ex6(int[] initial, int[] desired) {
+		this.initial = Arrays.copyOf(initial, initial.length);
+		this.desired = Arrays.copyOf(desired, desired.length);
+		moves = moveCars();
+	}
+	
+	/**
+	 * Print all the moves used in the program by order of occurrence
+	 * @param moves List of moves
+	 */
+	public void printMoves() {
+		if (this.moves.size() == 0) {
+			System.out.println("no need to move.");
+			return;
+		}
+		
+		this.moves.forEach(System.out::println);
+	}
+	
+	/**
+	 * Move cars in the parking lot from initial state to desired state
+	 * @return A list of Move objects representing the moves done
+	 */
+	public ArrayList<Move> moveCars() {
+		//a map between cars and their current position
+		HashMap<Integer, Integer> parkingLot = new HashMap<>();
+		
+		//list of moves to be returned
+		ArrayList<Move> moves = new ArrayList<Move>();
+		
+		//initialize map with cars which are not already in place
+		for (int i = 0; i < initial.length; i++) {
+			if (initial[i] == 0 || (initial[i] != 0 && initial[i] != desired[i])) {
```

Comment:
I think that condition initial[i] != 0 is not necessary as it is going there only if initial[i] is not 0 (the second statement of "or" is going to be evaluated only if the first one is false)